tests/tests/loopback.rs

```diff
@@ -0,0 +1,306 @@
+            target: RouterTarget::Ip(dest.parse().unwrap()),
+        })?;
+        Ok(())
+    }
+
+    /// Allow all traffic through the overlay firewall.
+    fn fw_allow_all(&self) -> Result<()> {
+        let adm = OpteAdm::open(OpteAdm::XDE_CTL)?;
+        let mut filters = Filters::new();
+        filters.set_hosts(Address::Any);
+        filters.set_ports(Ports::Any);
+        adm.add_firewall_rule(
+            &self.name,
+            &FirewallRule {
+                direction: Direction::In,
+                action: FirewallAction::Allow,
+                priority: 0,
+                filters,
+            },
+        )?;
+        Ok(())
+    }
+
+    /// Return the guest mac as an array of bytes.
+    fn mac(&self) -> [u8; 6] {
+        self.cfg.guest_mac.bytes()
+    }
+
+    /// Return the guest IP address as a string.
+    fn ip(&self) -> String {
+        match &self.cfg.ip_cfg {
+            IpCfg::Ipv4(cfg) => cfg.private_ip.to_string(),
+            _ => panic!("expected ipv4 gueset"),
+        }
+    }
+
+    /// Return the source underlay address.
+    fn underlay_ip(&self) -> std::net::Ipv6Addr {
+        self.cfg.phys_ip.into()
+    }
+}
+
+impl Drop for OptePort {
+    /// When this port is dropped, remove it from the underlying xde device.
+    fn drop(&mut self) {
+        let adm = match OpteAdm::open(OpteAdm::XDE_CTL) {
+            Ok(adm) => adm,
+            Err(e) => {
+                eprintln!("failed to open xde device on drop: {}", e);
+                return;
+            }
+        };
+        if let Err(e) = adm.delete_xde(&self.name) {
+            eprintln!("failed to delete xde on drop: {}", e);
+        }
+    }
+}
+
+/// This is resource handle for an xde device. It provides a few convenience
+/// methods for setting up global OPTE properties. It also removes the xde
+/// driver from the kernel when dropped. This is helpful for cleaning things up
+/// after a test run.
+struct Xde {}
+
+impl Xde {
+    /// Set the underlay data links that all OPTE ports will use.
+    fn set_xde_underlay(dev0: &str, dev1: &str) -> Result<()> {
+        let adm = OpteAdm::open(OpteAdm::XDE_CTL)?;
+        adm.set_xde_underlay(dev0, dev1)?;
+        Ok(())
+    }
+
+    /// Set the virtual to physical port mappings that all OPTE ports will use.
+    fn set_v2p(vip: &str, ether: &str, ip: &str) -> Result<()> {
+        let adm = OpteAdm::open(OpteAdm::XDE_CTL)?;
+        adm.set_v2p(&SetVirt2PhysReq {
+            vip: vip.parse().unwrap(),
+            phys: PhysNet {
+                ether: ether.parse().unwrap(),
+                ip: ip.parse().unwrap(),
+                vni: Vni::new(1701u32).unwrap(),
+            },
+        })?;
+        Ok(())
+    }
+}
+impl Drop for Xde {
+    /// When this object is dropped, remove the xde kernel module from the
+    /// underlying system.
+    fn drop(&mut self) {
+        let mut cmd = Command::new("pfexec");
+        cmd.args(["rem_drv", "xde"]);
```

Comment:
Why are we removing the driver? It just seems a bit heavy-handed, especially if someone is running this test locally.

Reply:
See later comment.

Reply:
Perhaps an ENV variable to allow overriding this?